src/forum.cpp

```diff
@@ -1,147 +1,235 @@
+
+    statuses status_table(_self, _self);
+
+    if (content.size() == 0) {
+        auto& row = status_table.get(account, "no previous status entry for this account");
+        status_table.erase(row);
+    } else {
+        update_status(status_table, account, [&](auto& row) { 
+            row.content = content;
+        });
+    }
+}
+
+// @abi
+void forum::vote(
+    const account_name voter, 
+    const account_name proposer, 
+    const name proposal_name, 
+    const std::string& proposal_hash, 
+    uint8_t vote, 
+    const std::string& vote_json
+) {
+    require_auth(voter);
+
+    proposals proposal_table(_self, proposer);
+    proposal_table.get(proposal_name, "proposal_name does not exist under proposer's scope");
+
+    // The proposal_hash should be a hash of "title" + the JSON in the `proposal_json` field, appended directly.
+    // TODO: check that here.
+    eosio_assert(proposal_hash.size() < 128, "proposal_hash should be less than 128 characters long.");
+
+    VALIDATE_JSON(vote_json, 8192);
+
+    votes vote_table(_self, proposer);
+    update_vote(vote_table, proposal_name, voter, [&](auto& row) {
+        row.vote = vote;
+        row.vote_json = vote_json;
+    });
+}
+
+// @abi
+void forum::unvote(
+    const account_name voter,
+    const account_name proposer,
+    const name proposal_name,
+    const std::string& proposal_hash
+) {
+    require_auth(voter);
+
+    proposals proposal_table(_self, proposer);
+    proposal_table.get(proposal_name, "proposal_name does not exist under proposer's scope");
+
+    // The proposal_hash should be a hash of "title" + the JSON in the `proposal_json` field, appended directly.
+    // TODO: check that here.
+    eosio_assert(proposal_hash.size() < 128, "proposal_hash should be less than 128 characters long.");
+
+    votes vote_table(_self, proposer);
+    
+    auto index = vote_table.template get_index<N(votekey)>();
+    auto vote_key = compute_vote_key(proposal_name, voter);
+
+    auto itr = index.find(vote_key);
+    eosio_assert(itr != index.end(), "no vote exists for this proposal_name/voter pair");
+    
+    vote_table.erase(*itr);
+}
+
+void forum::cleanvotes(
+    const account_name proposer,
+    const name proposal_name,
+    uint64_t max_count
+) {
+    proposals proposal_table(_self, proposer);
+
+    auto itr = proposal_table.find(proposal_name);
+    eosio_assert(itr == proposal_table.end(), "proposal_name must not exist anymore");
+
+    votes vote_table(_self, proposer);
+    auto index = vote_table.template get_index<N(votekey)>();
+
+    auto vote_key_lower_bound = compute_vote_key(proposal_name, 0x0000000000000000);
+    auto vote_key_upper_bound = compute_vote_key(proposal_name, 0xFFFFFFFFFFFFFFFF);
+
+    auto lower_itr = index.lower_bound(vote_key_lower_bound);
+    auto upper_itr = index.upper_bound(vote_key_upper_bound);
+
+    uint64_t count = 0;
+    while (count < max_count && lower_itr != upper_itr) {
```

Comment:
would that prevent user `0xfffffffffffffffffffffffffff` to vote ?

Reply:
Need more testing, but I don't see why it would prevent voting, could prevent cleaning correctly though. Will need to check if bound are inclusive or not.